### ISSUE TITLE
training and testing with tmtoolkit

### ISSUE THREAD

**Issue by ihavemanyquestions** [WZBSocialScienceCenter/tmtoolkit#18]
Hi, 

First of all thank you for your work on tmtoolkit. It's by far the best and most complete package for topic modelling I came across. 

I have a question which causes me a great deal of confusion. I am trying to train an LDA model and fit it on a train dataset. And then evaluate a test set using evaluate_topic_models using the code below: 
`
        tf_train = tf_vectorizer.fit_transform(data_samples)
        tf_test = tf_vectorizer.transform(test_samples)
   
        t0 = time()
            
        var_params = [{'n_topics': k} for k in range(1, 51, 1)]
        const_params = {'n_iter': 100,'random_state': 100}
            
        training = compute_models_parallel(**tf_train**,varying_parameters=var_params,constant_parameters=const_params, n_max_processes=multip.cpu_count())
        
        eval_results = evaluate_topic_models(**tf_test**,varying_parameters=var_params,
                                                                       constant_parameters=const_params,
                                                                       return_models=False,
                                                                       n_max_processes=multip.cpu_count(),
                                                                       metric=)`
    
Now, I suspect this code is probably wrong because there is no obvious link between how the trained&fitted model is the one used for evaluation. **Is the tmtoolkit package purely generative? Can it be used in a train/test fashion at all?** I have read the documentation repeatedly and I feel a bit lost. 

I would appreciate any help trying to understand this and I think it would also be useful for other people who come across this question and try to use the package in the same way as I do. 

PS: here is a MATLAB implementation example which uses the train/test approach to generate perplexity on a 'validation' set of documents other than the one the model was trained on: [(https://uk.mathworks.com/help/textanalytics/ug/choose-number-of-topics-for-LDA-model.html)]

**Comment by internaut** [comment]
Hi,

if I understand correctly, you want to calculate some kind of model fit using held out documents in your test data in order to evaluate your topic model. There's already an evaluation method that uses a similar approach, called [`metric_held_out_documents_wallach09()`](https://tmtoolkit.readthedocs.io/en/latest/api.html#tmtoolkit.topicmod.evaluate.metric_held_out_documents_wallach09).

If you want to implement your own evaluation method, you can check out the source code in `topicmod.tm_...` and `topicmod.parallel` to see how it's done. There's currently no "easy" way to provide your own evaluation method. tmtoolkit is focused on users that apply the already implemented evaluation methods.

**Comment by ihavemanyquestions** [comment]
Hi, 

Thank you for answering quickly. Then am I correct to say that metric_held_out_documents_wallach09() does the evaluation of the 'held out' documents without having fitted the LDA itself on them, but on the non-held out documents? 
(sorry I realise this might be straightforwards for people more experienced with LDA methods). 

Thank you!

**Comment by internaut** [comment]
Don't know if I understand correctly. Held out documents are *held out* so they're not used when fitting the LDA model. Please refer to the code here: https://github.com/WZBSocialScienceCenter/tmtoolkit/blob/master/tmtoolkit/topicmod/tm_sklearn.py#L142

**Comment by ihavemanyquestions** [comment]
Hi, after studying the tmtoolkit code in more depth i am coming back to this (**non**)issue (it seems it was  just my confusion, so i'll close it), to mention some things that I think would be very useful to be added to tmtoolkit if more additions are in plan:

- ability to choose the proportion of the held out documents as a varying parameter
- ability to choose the number of folds for validation as a varying parameter
- ability to calculate coherence measures for held out documents
